ranger/data/scope.sh

```diff
@@ -77,13 +76,39 @@ handle_extension() {
             odt2txt "${FILE_PATH}" && exit 5
             exit 1;;
 
+	# RTF and DOC
+	rtf|doc)
+	    # Preview as text conversion
+	    # note: catdoc does not always work for .doc files
+	    catdoc "${FILE_PATH}" && exit 5
+	    exit 1;;
+
+	# DOCX
+	docx)
```

Comment:
Can't pandoc handle rtf and doc? Maybe we can add pandoc after catdoc? Check the pdf handling with mutool and pdftotext to see how we implement a fallback.

Reply:
No, it cannot. Here again, despite the similar names, they are very different types files.

Reply:
> Check the pdf handling with mutool and pdftotext to see how we implement a fallback.

pdftotext does not handle .docx files. I don't have mutool, but according to its man:

`The supported input document formats are: pdf, xps, cbz, and epub.`

Reply:
All the applications I mention are quick and work reliably, except for catdoc with .doc. It should work and does work for some files, but not for others. It may be because I am testing it on .doc files which I create with LibreOffice rather than MS Word. But that is not a big deal since .doc and .xls are archaic formats that very few people use now that they have been replaced with their "x" version years ago.

Reply:
All these MS formats are a total nightmare anyway :laughing: and I doubt ranger users love them. But unfortunately we sometimes have to use them when working with the rest of the world :expressionless:

Reply:
You're right, pandoc only handles rtf as an output format. Maybe we should add all the input formats pandoc accepts to the case though? Not really a reason not to support all the formats the tool supports I think.

You misunderstand what I said about the pdf previews. Those tools only do pdf to plaintext conversion. What you should have a look at is how both those programs are called from the same case. This is useful when there's multiple programs that could handle a single format, then you can add one that's unlikely to be installed at the top and if it's not installed scope'll fall back to the next one.

Reply:
> You're right, pandoc only handles rtf as an output format.

Well, the problem here is that pandoc does not accept .doc nor .rtf as an input format. It does however accept .docx

Reply:
> You misunderstand what I said about the pdf previews. Those tools only do pdf to plaintext conversion. What you should have a look at is how both those programs are called from the same case. This is useful when there's multiple programs that could handle a single format, then you can add one that's unlikely to be installed at the top and if it's not installed scope'll fall back to the next one.

I think that we are both misunderstanding each other: there is no fallback for .rtf and .doc files.

There are other tools, besides catdoc, which can handle those files (unoconv for instance), but as far as I know, they cannot output to the terminal: they can only create a new file. So they did not seem suitable for the preview.

Reply:
Having a temp text file created by unoconv and then pipe it to get the preview seemed heavy and twisted. Particularly since catdoc works flawlessly for .rtf files and .doc files, for which it is a bit hit or miss, is a disappearing format.

Reply:
> Maybe we should add all the input formats pandoc accepts to the case though? Not really a reason not to support all the formats the tool supports I think.

I'll have a look at that.

Reply:
I was thinking of having pandoc be a fallback for all the formats it supports.

Reply:
One of the formats that pandoc accepts is .odt and there is already a tool (odt2txt) in scope.sh (and it is not commented).

If I add .odt to the pandoc paragraph, I don't know what would happen if both are uncommented. Will this create a conflict? If so, then I could comment the pandoc paragraph, but it is a bummer for all the other formats.

Reply:
Of course, the odt2txt paragraph could be commented instead, but pandoc is a much heavier program and having it as the default for .odt files may not be the right decision? People can of course customize their scope.sh.

Or maybe it doesn't matter if both paragraphs are uncommented??? I just wonder what would happen if someone has both programs.

Reply:
A lot of pandoc input formats are markup languages. So those would be irrelevant (no point converting from one markdown format to another for instance).

And some of the formats are already covered elsewhere in scope.sh (odt2txt I mentioned earlier, but also .html, .json, etc.).

And finally I am totally unfamiliar with some of the formats and I don't know whether they are text (so irrelevant here) or not.

So, I am not comfortable to add more to pandoc. If someone wants to do it of course, feel free to jump in.

Reply:
Things like this is exactly what I described the fallback for, look at mutool/pdftotext to see how you can handle this. (Newer bash syntax has support for matching several clauses but let's not stray *further* from posix sh.)

Reply:
Relevant formats I'd say are docx, epub, fb2 and odt. You're right that most of the markup languages are reasonably readable and people may not want to have them be rendered.

Reply:
> Things like this is exactly what I described the fallback for, look at mutool/pdftotext to see how you can handle this. (Newer bash syntax has support for matching several clauses but let's not stray further from posix sh.)

I understand. I just wasn't sure I had to add a pandoc line for all of those already covered formats (since pandoc overlaps with several tools).

Reply:
Yes, that's not very satisfying. We *could* merge all those cases and use the fallthroughs, it'd probably work but that seems worse than the duplication tbh. Another alternative is wrapping pandoc in a function but I'm not a fan of that either.

Reply:
Well, because of the multi-overlaps of tools, it is unclear to me how to write it:

Currently, there is:

```sh
        # OpenDocument
        odt|ods|odp|sxw)
            # Preview as text conversion
            odt2txt "${FILE_PATH}" && exit 5
            exit 1;;
```

I can't add pandoc here because pandoc does not accept .ods, .odp, .sxw.

I could take .odt out of this and put it elsewhere, where I can add pandoc, but that changes the current default for .odt files.

Seems that the only way out is to add some commented paragraph.

Reply:
And if I add .odt with .docx (which would make sense), then I have to comment that out and that kills the option for .docx unless the scope.sh is edited.

That's what I was trying to explain. It gets a bit messy and complicated.

I will add .odt by itself with pandoc and comment it out.

Reply:
> We could merge all those cases and use the fallthroughs

The complication is that pandoc overlaps several categories, but does not cover all the subcategories in those categories.

Anyhow, I will submit something in 2 min and you can re-asses.

Reply:
> Another alternative is wrapping pandoc in a function but I'm not a fan of that either.

I guess, that would be the clean way to do this. But might be overkill for this purpose.

Reply:
The `odt|ods|odp...` isn't an issue. If odt2txt fails, it falls through to pandoc, if that fails no preview is shown. I think that's what the `exit 5` means (it's at the top of `scope.sh`).

Reply:
Something inspired by rifle might be best :slightly_smiling_face:

Reply:
Oh, I see.

Reply:
OK, that simplify things already. I'll change this.

Reply:
Well, I guess, instead of putting fb2 and epub with docx as I did, I could add a separate paragraph for them with pandoc next to where they were initially covered. That would be much less messy.

Reply:
I'll change that too.

Reply:
Yes, keep all the original clauses, just add the fall through with pandoc.

Reply:
I think the current version is more sensible than my first attempts at adding more pandoc input formats. How does it look?

Reply:
Note though that I haven't tested pandoc with the new formats I added... (I don't have any files with those formats).

So I don't know whether it works and whether it adds a delay.

Reply:
If you have epub and fb2 files and want to test it out, that would be best.